### ISSUE TITLE
avoid_shadowing_type_parameters does not catch generic function types in new places

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#2958]
**Describe the issue**

As of Dart 2.14 (I think), you can use a generic function type (e.g. `void Function<T>(T)`) as a type argument and as a bound. This opens up new opportunities for shadowing! Let's take a look:

```dart
class F<T extends void Function<T>(T)> {} // LINT

abstract class F<T> {
  void Function<T>(T) get fn; // LINT
  List<void Function<T>(T)> get list; // LINT
}

extension <T> on void Function<T>(T) {} // LINT
```

(Some of these were valid code before 2.14, I think. Either way.)

None of these four positions are reported as lint, but they should be.

**To Reproduce**

Throw those examples in the test file.

**Expected behavior**

4 new lint should be reported.

**Comment by pq** [comment]
Those examples are pretty brain bending. 🤯 

Thanks for reporting!
